pixeltable/catalog/catalog.py

```diff
@@ -422,6 +425,176 @@ def create_view(
+        # Only the table being replicated is actually made visible in the catalog.
+        self._tbls[tbl_id] = self._load_tbl(tbl_id)
+        return self._tbls[tbl_id]
+
+    def __store_replica_md(self, path: Path, md: schema.FullTableMd) -> None:
+        _logger.info(f'Creating replica table at {path!r} with ID: {md.tbl_md.tbl_id}')
+        dir = self._get_schema_object(path.parent, expected=Dir, raise_if_not_exists=True)
+        assert dir is not None
+
+        conn = Env.get().conn
+        tbl_id = md.tbl_md.tbl_id
+
+        new_tbl_md: Optional[schema.TableMd] = None
+        new_version_md: Optional[schema.TableVersionMd] = None
+        new_schema_version_md: Optional[schema.TableSchemaVersionMd] = None
+
+        # We need to ensure that the table metadata in the catalog always reflects the latest observed version of
+        # this table. (In particular, if this is a base table, then its table metadata need to be consistent
+        # with the latest version of this table having a replicated view somewhere in the catalog.)
+        q: sql.Executable = sql.select(schema.Table.md).where(schema.Table.id == tbl_id)
+        existing_md_row = conn.execute(q).one_or_none()
+
+        if existing_md_row is None:
+            # No existing table, so create a new record.
+            # If this is a proper ancestor of the table being replicated, it might be the case that the current_version
+            # and/or current_schema_version records in the given TableMd are strictly greater than the versions found
+            # in the given TableVersionMd and/or TableSchemaVersionMd. If that's the case, we need to adjust them so
+            # that they point to known versions. This will cause no harm, since those proper ancestors will not be
+            # directly instantiable.
+            q = sql.insert(schema.Table.__table__).values(
+                id=tbl_id,
+                dir_id=dir._id,
+                md=dataclasses.asdict(
+                    dataclasses.replace(
+                        md.tbl_md,
+                        name=path.name,
+                        user=Env.get().user,
+                        is_replica=True,
+                        current_version=md.version_md.version,
+                        current_schema_version=md.schema_version_md.schema_version,
+                    )
+                ),
+            )
+            conn.execute(q)
+        elif md.tbl_md.current_version > existing_md_row.md['current_version']:
+            # New metadata is more recent than the metadata currently stored in the DB; we'll update the record
+            # in place in the DB.
+            # The same consideration applies to current_version etc. as before, except now we also need to account for
+            # other versions already in the DB, which might be more recent.
+            new_current_version = max(existing_md_row.md['current_version'], md.version_md.version)
```

Comment:
At this point I'm finding it hard to reason about how md.tbl_md.current_version is related to md.version_md.version.

It would make things easier if you stipulated that for replica md, tbl_md.current_version == version_md.version (and the same for current_schema_version). Or you set current_version/current_schema_version to None.